### PR TITLE
APPS: simplify `x509` cert generation; allow pubkey input options to read private key

### PR DESCRIPTION
**Pull request by DDvO**
When reviewing the addition to `test/smime-certs/mksmime-certs.sh` by #18567,
I noticed again that producing certs using `openssl x509` is still needlessly cumbersome.
While there are already abbreviations for the case of self-issued (including self-signed) certs,
for other certs (produced with the `-CA` option) one is so far forced to generate a cert request first.
It turns out this was simply because is was not possible to combine the `-key` option with `-CA`.

**Update:** Due to the comments by @vdukhovni, changed the solution to allow reading a private key when a public key is expected. This enables conveniently using the `-force_pubkey` option with `-CA`.

This PR adds the missing feature and makes use of the simplification in `mksmime-certs.sh`.
It also adds a test case for the new feature in `25-test_x509.t`.

On this occasion, also clean up the key input options of other apps such as `pkey`.

On this occasion it also cleans up `mksmime-certs.sh`, speeds up its cert generation by not freshly generating RSA and DSA keys, and fixes its generation of inconsistent key usages for non-RSA certs.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Ping @openssl/committers for 2nd review.

File: doc/man1/openssl-x509.pod.in
Comment (tmshort) [inline doc/man1/openssl-x509.pod.in:168]: "usually" is a wishy-washy word. Could this be re-phrased?

**Review by tmshort — APPROVED** [review]
Other than a wishy-washy word in the documentation, looks ok to me.

File: doc/man1/openssl-x509.pod.in
Comment (DDvO) [reply doc/man1/openssl-x509.pod.in]: Thanks  @tmshort for your comment and prospective approval.
I've improved the wording, no more using "usually".

Can you and @t8m please re-approve.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@tmshort please reconfirm

**Comment by tmshort** [comment]
reconfirm

**Comment by vdukhovni** [comment]
With `-CA`, it seems that the only part of the `-key` file that is used is the public component.  If so, it seems that this should support (or require) a **public** key, rather than a private key.  After all, if a CA is signing a certificate, why should the subject's *private* key be required???

**Comment by beldmit** [comment]
So don't we get a mixed semantics for the `-key` option in case when your patch is applied?

**Comment by vdukhovni** [comment]
Ideally, the `-key` option would load either a private key or public key whichever is provided, and then extract the public part if a private key is provided.  In other words, make the "user experience" as seamless as possible.

In any case, think about what is the least surprising (most natural) interface, and do that.  Sure, this could be a followup PR, if you don't want to reopen reviews of the work so far.  I'd be tempted to just amend this one, it is not a large PR.  Your (and the other reviewers') call.

**Comment by DDvO** [comment]
> So don't we get a mixed semantics for the `-key` option in case when your patch is applied?

With the current state of the PR, the `-key` argument must be a private key file.

> Ideally, the `-key` option would load either a private key or public key whichever is provided, and then extract the public part if a private key is provided. In other words, make the "user experience" as seamless as possible.

Interesting idea.
Is there already any other `openssl` command option that supports reading a file that contains a public _or_ private key?

> In any case, think about what is the least surprising (most natural) interface, and do that. Sure, this could be a followup PR, if you don't want to reopen reviews of the work so far. I'd be tempted to just amend this one, it is not a large PR. Your (and the other reviewers') call.

I was concerned exactly with this point, since the approval is already done.
@t8m and @tmshort, what do you think?

**Comment by vdukhovni** [comment]
> Interesting idea.
> Is there already any other `openssl` command option that supports reading a file that contains a public _or_ private key?

I can't think of one offhand, but in principle any command that expects a **public** key should be willing to read a private key instead, and extract the public key from that.  That would be a general UI improvement.

**Comment by vdukhovni** [comment]
In this case, the option's expected data would depend on other options (`-CA` changes the semantics of `-key`).  Is that natural or surprising? Should some other option actually be used, leaving `-key` as incompatible with `-CA`?

These questions perhaps mean that the PR should not immediately be merged.  We should at least decide what the UI should look like long-term, and then perhaps merge this PR, with other PRs to follow, or amend it as seems best.

One way or another, when all we want is a public key, it should be possible to specify a file with just the public key.  Whether that's a separate public-key-only option, or a combo option that sometimes wants a private key, and other times a public key.

Please think through what a user might expect in terms of behaviour, and what can be most clearly explained in the documentation, ...

**Comment by DDvO** [comment]
Thanks @vdukhovni for your helpful comments.
This way I found a solution that is straightforward to implement, use, and describe for `x509`
and also improves the public key input options of other apps likey `pkey`:
Whenever a public key is expected/sufficient, allow as input also a private key.

Thus, a CA-issued cert can now be produced as simply as this:
```
openssl x509 -new -force_pubkey priv-or-pub-key.pem -subj "/CN=any" -CA ca-cert+key.pem
```

**Comment by vdukhovni** [comment]
> Thanks @vdukhovni for your helpful comments. This way I found a solution that is straightforward to implement, use, and describe for `x509` and also improves the public key input options of other apps likey `pkey`: Whenever a public key is expected/sufficient, allow as input also a private key.
> 
> Thus, a CA-issued cert can now be produced as simply as this:
> 
> ```
> openssl x509 -new -force_pubkey priv-or-pub-key.pem -subj "/CN=any" -CA ca-cert+key.pem
> ```

Excellent, I am glad that a bit of out-of-the-box thinking has produced a cleaner interface and improved the usability of pubkey input across the board.

A couple of decades of working on Postfix with Wietse has taught me to try to think beyond the immediate symptoms and look for opportunities to solve the general problem as cleanly as possible.  Always look for interfaces that are easy to explain and apply broadly and are likely to be maintainable long-term.

File: apps/dsa.c
Comment (vdukhovni) [inline apps/dsa.c:170]:
```suggestion
    private = !pubin && (!pubout || text);
```

File: apps/dsa.c
Comment (vdukhovni) [inline apps/dsa.c:243]: No need to change the order here or immediately below.

File: apps/ec.c
Comment (vdukhovni) [inline apps/ec.c:166]: See above.

File: apps/lib/apps.c
Comment (vdukhovni) [inline apps/lib/apps.c:586]: No need to make this change.

File: apps/lib/apps.c
Comment (vdukhovni) [inline apps/lib/apps.c:612]: Is the error stack relevant here?  If the public key and private load fails, what will be reported?

Another way to handle this would be:

* Add another argument to `load_key_certs_crls` to request *quiet* operation, with no error output and the error stack cleared.
* Try to load a private key first, with `quiet=1`
* Finally, if that fails, try load a public key with `quiet=0`.

This avoids the kludgey messing around with bio_err...

File: apps/pkey.c
Comment (vdukhovni) [inline apps/pkey.c:191]: See above.

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:73]:
```suggestion
    {"pubin", OPT_PUBIN, '-', "Input key is a public key"},
```

File: apps/rsa.c
Comment (vdukhovni) [inline apps/rsa.c:226]: See above.

File: apps/rsa.c
Comment (vdukhovni) [inline apps/rsa.c:336]: No need here and below.

File: apps/rsautl.c
Comment (vdukhovni) [inline apps/rsautl.c:52]: See above.

File: doc/man1/openssl-dsa.pod.in
Comment (vdukhovni) [inline doc/man1/openssl-dsa.pod.in:121]:
```suggestion
By default, a private key is read from the input.
With this option a public key is read instead.
If the input is a private key, its public part is used.
```
Here and below...

File: doc/man1/openssl-x509.pod.in
Comment (vdukhovni) [inline doc/man1/openssl-x509.pod.in:411]: See above.

**Review by vdukhovni — CHANGES_REQUESTED** [review]

**Comment by vdukhovni** [comment]
@levitte Your thoughts on this PR?  Are there any reasons I may have missed that would make it problematic to try to load a private key as a fallback when loading a public key fails?  In what order should the attempts happen (I haven't through through questions of password prompting for private keys.  Would the password prompt happen just-in-time, or even before the content is recognised as an encrypted private key, ...)

File: apps/dsa.c
Comment (DDvO) [reply apps/dsa.c]: Simplified as suggested.

File: apps/dsa.c
Comment (DDvO) [reply apps/dsa.c]: Reverted both. This was just for consistency with the same statement in related `apps/*.c` files.

File: apps/ec.c
Comment (DDvO) [reply apps/ec.c]: Changed accordingly to `private = !pubin && (text || (!param_out && !pubout))`

File: apps/lib/apps.c
Comment (DDvO) [reply apps/lib/apps.c:586]: This improves adherence to the coding style and 
makes `check-format.pl apps/lib/apps.c` happy :smile:

File: apps/pkey.c
Comment (DDvO) [reply apps/pkey.c]: Reverted.

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c]: Done

File: apps/rsa.c
Comment (DDvO) [reply apps/rsa.c:226]: The ` ? 1 : 0` is just needless and potentially confusing.
I really think it's better just dropping it.

File: apps/rsa.c
Comment (DDvO) [reply apps/rsa.c]: Reverted both.

File: apps/rsautl.c
Comment (DDvO) [reply apps/rsautl.c]: Changed as requested to `Input key`.

File: doc/man1/openssl-dsa.pod.in
Comment (DDvO) [reply doc/man1/openssl-dsa.pod.in]: Changed the last sentence here and in all other 6 doc files to:
```
If the input contains no public key but a private key, its public part is used.
```
I kept the "no public key but a private key" because it is more precise.
The point is that an input file may contain multiple keys, and private keys are considered only if it does not contain any public key.

File: apps/lib/apps.c
Comment (DDvO) [reply apps/lib/apps.c:609]: Good point - done.

On this occasion spotted and removed a leftover outdated decl in `apps.h`:
```
int load_key_cert_crl(const char *uri, int format, int maybe_stdin,
                      const char *pass, const char *desc,
                      EVP_PKEY **ppkey, EVP_PKEY **ppubkey,
                      X509 **pcert, X509_CRL **pcrl);
```
and made `check-format.pl` happy on `apps/include/apps.h`.

**Comment by DDvO** [comment]
BTW, this PR unearths again the pain that there is horrible redundancy/overlap between `{rsa,pkey}utl.c` and among `{dsa,ec,rsa,pkey}.c}` and also within their docs.

**Review by vdukhovni — COMMENTED** [review]
Thanks for all the fixups. This resolves my concerns.

Given that many of the suggestions are mine, I think this now needs independent approvals.

File: apps/include/apps.h
Comment (beldmit) [inline apps/include/apps.h:71]: Are all the changes in this file codestyle-related? Can't we don't introduce them or introduce them separately?

**Review by beldmit — APPROVED** [review]
LGTM, though I don't like mixing style things with new features.

File: apps/include/apps.h
Comment (DDvO) [reply apps/include/apps.h:71]: The indentation if preprocessor directives according to the `#if` nesting level is part of the coding style and checked by the tool.
Before these changes, `check-format.pl` complained like this:
````
apps/include/apps.h:69:'#if' nesting indent = 0 != 1:#define app_load_config(filename) app_load_config_internal(filename, 0)
apps/include/apps.h:70:'#if' nesting indent = 0 != 1:#define app_load_config_quiet(filename) app_load_config_internal(filename, 1)
```

**Comment by DDvO** [comment]
Rebased to fix merge conflicts.

@t8m can you reapprove?

**Comment by DDvO** [comment]
There is an unrelated CI failure for `doc-nits`:
```
/usr/bin/perl ./util/find-doc-nits -c -n -l -e
doc/man3/CMS_add0_cert.pod:1: Section missing in CMS_sign_ex()
doc/man3/CMS_add0_cert.pod:1: Section missing in CMS_sign()
doc/man3/CMS_add0_cert.pod:1: Section missing in CMS_verify()
doc/man3/CMS_add0_cert.pod:1: Section missing in CMS_verify()
```
is unrelated; fix for that is in #20369.

**Comment by DDvO** [comment]
Rebased again, just to get rid of the above unrelated CI failure message.

**Review by t8m — APPROVED** [review]

**Comment by beldmit** [comment]
LGTM

**Comment by DDvO** [comment]
> LGTM

Thank you @beldmit for this statement.
Can you please make it formal and accordingly set **approval:done** ?

**Comment by beldmit** [comment]
Sorry, missed it

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by beldmit** [comment]
@t8m done

**Comment by DDvO** [comment]
Merged - thanks @t8m, @tmshort, @vdukhovni, and @beldmit for your reviews and approvals.

**Pull request closed without merging**
